README.md
# JOYSO on Tron
JOYSO on Tron API client library for trading.

## Installation
You can use this command to install:

    npm install tron-joyso

## Usage
Setup and connect to JOYSO
```JavaScript
const Joyso = require('tron-joyso');

async function start() {
  const joyso = new Joyso({
    // your private key
    key: '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef'
  });

  await joyso.connect();
}

```

### subscribeOrderBook(pair, callback)
Subscribe order book, notify if change.
```JavaScript
const subscription = joyso.subscribeOrderBook('JOY_TRX', orderBook => {
  console.log(JSON.stringify(orderBook));
});
```
Result:
```JSON
{
  "buy":[
    {
      "price":1.23456,
      "amount":"8.5"
    },
    {
      "price":1.23455,
      "amount":"98.5"
    }
  ],
  "sell":[
    {
      "price":1.2346,
      "amount":"100"
    },
    {
      "price":1.2347,
      "amount":"500"
    }
  ]
}
```
* amount is BigNumber object.

### subscribeTrades(pair, callback)
Subscribe market trades, notify if change, return last 100 records.
```JavaScript
const subscription = joyso.subscribeTrades('JOY_TRX', trades => {
  console.log(JSON.stringify(trades.slice(0, 2)));
});
```
Result
```JSON
[
  {
    "id":317,
    "side":"sell",
    "price":1.23456,
    "amount":"2",
    "pair":"JOY_TRX"
  },
  {
    "id":315,
    "side":"buy",
    "price":1.2347,
    "amount":"1",
    "pair":"JOY_TRX"
  }
]
```
* amount is BigNumber object.

### subscribeBalances(callback)
Subscribe balances, notify if change.
```JavaScript
const subscription = joyso.subscribeBalances(balances => {
  console.log(JSON.stringify(balances));
});
```
Result
```JSON
{
  "JOY":{
    "inOrder":"0",
    "available":"4.5"
  },
  "TRX":{
    "inOrder":"144.7757819",
    "available":"97815.997145"
  }
}
```
* inOrder and available are BigNumber objects.

### subscribeOrders(callback)
Subscribe open orders, notify if change.
```JavaScript
const subscription = joyso.subscribeOrders(orders => {
  console.log(JSON.stringify(orders));
});
```
Result
```JSON
[
  {
    "id":353,
    "status":"active",
    "side":"buy",
    "price":12.3481,
    "amount":"1",
    "fill":"0",
    "pair":"T00_TRX"
  },
  {
    "id":326,
    "status":"partial",
    "side":"buy",
    "price":1.23456,
    "amount":"12",
    "fill":"3.5",
    "pair":"JOY_TRX"
  }
]
```
* amount and fill are BigNumber objects.
* status could be `active` or `partial`

### subscribeMyTrades(callback)
Subscribe my trades, notify if change, return last 100 records.
```JavaScript
const subscription = joyso.subscribeMyTrades(trades => {
  console.log(JSON.stringify(trades.slice(0, 2)));
});
```
Result
```JSON
[
  {
    "id":317,
    "status":"done",
    "txHash":"0xcf0aeb815200951559a38650a84f8eefa46411224e5e4076d6313ab47c7f9bb5",
    "side":"sell",
    "price":1.23456,
    "amount":"2",
    "pair":"JOY_TRX",
    "fee":"TRX",
    "gasFee":"0",
    "txFee":"0.002469"
  },
  {
    "id":316,
    "status":"done",
    "txHash":"0x582cc7a84e8aa7e28e44b11e22f24169a34776915ebbc95a88fa0e77c44faf4c",
    "side":"sell",
    "price":1.2347,
    "amount":"1",
    "pair":"JOY_TRX",
    "fee":"TRX",
    "gasFee":"0.000105",
    "txFee":"0.002469"
  }
]
```
* amount, gasFee and txFee are BigNumber objects.

### subscribeFunds(callback)
Subscribe funds, notify if change, return last 100 records.
```JavaScript
const subscription = joyso.subscribeFunds(funds => {
  console.log(JSON.stringify(funds));
});
```
Result
```JSON
[
  {
    "id":192,
    "status":"done",
    "txHash":"0x4dbc49ae4735b1c230244d41377cf6aeccd70c5181df048e3be8306af8a487e6",
    "type":"withdraw",
    "amount":"99",
    "token":"TRX",
    "fee":"TRX",
    "withdrawFee":"0.1",
    "timestamp":1537434044,
    "blockId":null
  },
  {
    "id":191,
    "status":"done",
    "txHash":"0x8435bf9f69dd908373d50353ebab343b625527cd8ea44532eb01c8b0a5642879",
    "type":"withdraw",
    "amount":"1",
    "token":"TRX",
    "fee":"JOY",
    "withdrawFee":"0.809841",
    "timestamp":1537433888,
    "blockId":null
  }
]
```
* amount and withdrawFee are BigNumber objects.
* status could be `pending`, `processing`, `done` or `failed`
* type could be `deposit`, `withdraw` or `transfer`

### buy({ pair, price, amount, feeByJoy })
Place buying order
```JavaScript
try {
  let order = await joyso.buy({
    pair: 'JOY_TRX',
    price: '1.23481',
    amount: 1,
    feeByJoy: true
  });
  console.log(JSON.stringify(order));
} catch (e) {
  if (e.statusCode === 400) {
    console.log(e.error.error);
  } else {
    console.log(e.message);
  }
}
```
Options

|Name|Required|Description|
|---|---|---|
|pair|O|Pair to trade, format is `${base}_${quote}`, eg: JOY_TRX|
|price|O|Order price, minimum is 0.000000001|
|amount|O|Quote amount|
|feeByJoy||Specify how to pay fee. `true` will pay by JOY. `false` will pay by quote token(TRX if pair XXX_TRX). Default is `false`|

Result
```JSON
{
  "id":361,
  "status":"complete",
  "side":"buy",
  "price":1.23481,
  "amount":"1",
  "fill":"1",
  "pair":"JOY_TRX"
}
```
* amount and fill are BigNumber objects.
* status could be `active`, `partial` or `complete`

### sell({ pair, price, amount, feeByJoy })
Place selling order
```JavaScript
let order = await joyso.sell({
  pair: 'JOY_TRX',
  price: '1.23481',
  amount: 100
});
```
Options and result are same with buy.

### trade({ pair, price, amount, feeByJoy, side })
Place order
```JavaScript
let order = await joyso.trade({
  side: 'buy',
  pair: 'JOY_TRX',
  price: '1.23481',
  amount: 100
});
```
Options and result are same with buy. One extra options

|Name|Required|Description|
|---|---|---|
|side|O|`buy` or `sell`|

### withdraw({ token, amount, fee })
Withdraw
```JavaScript
await joyso.withdraw({
  token: 'TRX',
  amount: 10,
  fee: 'trx'
});
```
Options

|Name|Required|Description|
|---|---|---|
|token|O|Token to withdraw|
|amount|O|Amount to withdraw|
|fee|O|Specify how to pay fee. `trx`, `joy` or `token`. `token` can only be used when token is quote token.|

### getMyTrades({ from, to, quote, base, side, before, limit })
Get my trades
```JavaScript
await joyso.getMyTrades({
  quote: 'TRX',
  base: 'JOY',
  side: 'sell',
  from: 1539129600,
  to: 1539216000,
  before: 123,
  limit: 10
});
```
Options

|Name|Required|Description|
|---|---|---|
|quote||Quote token|
|base||Base token|
|side||Specify side. `buy`, `sell` or blank. Blank means both.|
|from||From time (included). Unix timestamp|
|to||To time (excluded). Unix timestamp|
|before||Only return Trade ID before this. (excluded)|
|limit||Specify size of records to return. Default 100, max 1000|

Results are same with subscribeMyTrades.

### disconnect()
Disconnect from JOYSO.

### subscription.unsubscribe()
Unsubscribe
```JavaScript
subscription.unsubscribe();
```


## License
The project is released under the [MIT license](http://www.opensource.org/licenses/MIT).

## Contact
The project's website is located at https://github.com/Joyso-io/tron-joyso-api


CHANGELOG.md
# Change Log

## v0.4.1 / 2019-01-03
### Added
- create project


src/funds.js
const rp = require('request-promise');
const BigNumber = require('bignumber.js');
const _ = require('lodash');

const STATUS = ['pending', 'processing', 'done', 'failed', 'cancelled'];
const TYPE = ['deposit', 'withdraw', 'transfer'];

class Funds {
  constructor(options = {}) {
    this.client = options.client;
    this.address = options.address;
    this.onReceived = (() => {});
    this.funds = [];
    this.requesting = 0;
    this.requestId = 0;
  }

  subscribe() {
    this.cable = this.client.cable.subscriptions.create({
      channel: 'FundsChannel',
      contract: this.client.system.contract.substr(2),
      user: this.address.substr(2)
    }, {
      connected: () => {
        const index = _.findLastIndex(this.funds, fund => {
          return fund.status !== 'done' && fund.status !== 'failed';
        });
        if (index !== -1) {
          this.funds = this.funds.slice(index + 1);
        }
        this.update();
      },
      received: data => {
        switch (data.e) {
          case 'new':
            ++this.requestId;
            this.update();
            break;
          case 'update':
            const fund = this.funds.find(f => f.id === data.data.id);
            if (fund) {
              fund.status = STATUS[data.data.status];
              if (data.data.tx_hash) {
                fund.txHash = `0x${data.data.tx_hash}`;
              }
              this.onReceived(this.funds);
            } else {
              this.update();
            }
            break;
        }
      }
    });
  }

  unsubscribe() {
    clearTimeout(this.retryTimer);
    this.cable.unsubscribe();
    delete this.cable;
    this.balances = {};
  }

  async update() {
    if (this.requesting) {
      return;
    }
    this.requesting = this.requestId;
    clearTimeout(this.retryTimer);
    const after = this.funds.length ? this.funds[0].id : null;
    try {
      const json = await this.get(after);
      const funds = this.convert(json.funds, true);
      if (after) {
        this.funds.unshift(...funds);
      } else {
        this.funds = funds;
      }
      this.onReceived(this.funds);
      if (this.requesting !== this.requestId) {
        this.requesting = 0;
        this.update();
      } else {
        this.requesting = 0;
      }
    } catch (e) {
      console.log(e);
      this.requesting = 0;
      this.retryTimer = setTimeout(() => this.update(), 5000);
    }
  }

  convert(funds) {
    return funds.map(fund => {
      const token = this.client.tokenManager.addressMap[`0x${fund.token}`],
        amount = this.client.tokenManager.toAmount(token, fund.amount);
      let tokenFee, fee;
      if (fund.payment_method === 0) {
        tokenFee = this.client.tokenManager.trx;
        fee = 'TRX';
      } else if (fund.payment_method === 1) {
        tokenFee = this.client.tokenManager.joy;
        fee = 'JOY'
      } else {
        tokenFee = token;
        fee = token.symbol;
      }
      const withdrawFee = this.client.tokenManager.toAmount(tokenFee, fund.fee);
      return {
        id: fund.id,
        status: STATUS[fund.status],
        txHash: fund.tx_hash ? `0x${fund.tx_hash}` : null,
        type: TYPE[fund.type],
        amount: amount,
        token: token.symbol,
        fee: fee,
        withdrawFee: withdrawFee,
        timestamp: fund.timestamp,
        blockId: fund.block_id
      };
    });
  }

  get(after = null) {
    return this.client.request('funds', {
      data: {
        contract: this.client.system.contract.substr(2),
        user: this.address.substr(2),
        after: after
      }
    });
  }
}

module.exports = Funds;


samples/index.js
const Joyso = require('../src/joyso');

async function start() {
  const joyso = new Joyso({
    // your private key
    key: '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef'
  });

  await joyso.connect();

  // subscribe order book, notify if change
  joyso.subscribeOrderBook('JOY_TRX', orderBook => {
    console.log(JSON.stringify(orderBook));
  });

  // subscribe market trades, notify if change
  joyso.subscribeTrades('JOY_TRX', trades => {
    console.log(JSON.stringify(trades.slice(0, 5)));
  });

  // subscribe balances, notify if change
  joyso.subscribeBalances(balances => {
    console.log(JSON.stringify(balances));
  });

  // subscribe open orders, notify if change
  joyso.subscribeOrders(orders => {
    console.log(JSON.stringify(orders));
  });

  // subscribe my trades, notify if change
  joyso.subscribeMyTrades(trades => {
    console.log(JSON.stringify(trades));
  });

  // subscribe funds, notify if change
  joyso.subscribeFunds(funds => {
    console.log(JSON.stringify(funds));
  });

  try {
    let order;

    // place buying order
    order = await joyso.buy({
      pair: 'JOY_TRX',
      price: '1.1',
      amount: 1
    });

    // place selling order
    order = await joyso.sell({
      pair: 'JOY_TRX',
      price: '1.2',
      amount: 5
    });

    // or place order by trade with side
    order = await joyso.trade({
      side: 'buy',
      pair: 'JOY_TRX',
      price: '1.05',
      amount: 6,
      feeByJoy: true
    });

    // cancel order
    await joyso.cancel(2);

    // withdraw
    await joyso.withdraw({
      token: 'TRX',
      amount: 10,
      fee: 'trx'
    });

    // get my trades
    const trades = await joyso.getMyTrades({
      quote: 'ETH',
      base: 'JOY',
      side: 'sell',
      from: 1539680000,
      to: 1539679000,
      before: 123,
      limit: 10
    });
  } catch (e) {
    if (e.statusCode === 400) {
      console.log(e.error.error);
    } else {
      throw e;
      console.log(e.message);
    }
  }
}

start();

process.on('unhandledRejection', (reason, p) => {
  console.log('Unhandled Rejection:', reason);
  setTimeout(() => {
    process.exit(1);
  }, 1000);
});


src/token_manager.js
const BigNumber = require('bignumber.js');

class TokenManager {
  constructor(client, options) {
    this.client = client;
    this.pairs = {};
    this.reload(options);
  }

  subscribe() {
    this.cable = this.client.cable.subscriptions.create({
      channel: 'TokensChannel'
    }, {
      received: data => {
        return;
        switch (data.e) {
          case 'update':
            const token = this.addressMap[`0x${data.data.address}`];
            if (token) {
              if (data.data.price) {
                token.price = new BigNumber(data.data.price);
                token.gasFee = new BigNumber(data.data.gas_fee);
                token.withdrawFee = new BigNumber(data.data.withdraw_fee);
              } else {
                token.price = null;
                token.gasFee = null;
                token.withdrawFee = null;
              }
              token.depositable = data.data.depositable;
              token.withdrawable = data.data.withdrawable;
              token.tradable = data.data.tradable;
            }
            break;
        }
      }
    });
  }

  async refresh() {
    const json = await this.client.system.update();
    if (json) {
      this.reload(json);
    }
  }

  reload(json) {
    this.tokens = json.tokens;
    this.symbolMap = {};
    this.addressMap = {};
    json.tokens.forEach(t => {
      t.address = `0x${t.address}`;
      if (t.price) {
        t.price = new BigNumber(t.price);
        t.gasFee = new BigNumber(t.gas_fee);
        t.withdrawFee = new BigNumber(t.withdraw_fee);
      }
      this.symbolMap[t.symbol] = t;
      this.addressMap[t.address] = t;
    });
    this.trx = this.symbolMap.TRX;
    this.joy = this.symbolMap.JOY;
    this.quotes = json.quotes.map(t => {
      const token = this.addressMap[`0x${t.address}`];
      token.precision = t.precision;
      return token;
    });
    this.pairs = {};
    this.pairs['TRX'] = new Set(this.tokens.map(t => t.symbol));
    Object.keys(json.pairs).forEach(k => this.pairs[k] = new Set(json.pairs[k]));
    Object.keys(json.hidden_pairs).forEach(k => {
      const set = this.pairs[k];
      if (set) {
        json.hidden_pairs[k].forEach(t => set.delete(t));
      }
    });
  }

  getPair(pair) {
    let [base, quote] = pair.split('_');
    return this.pairs[quote] && this.pairs[quote].has(base) ? [this.symbolMap[base], this.symbolMap[quote]] : [];
  }

  toAmount(token, rawAmount) {
    if (rawAmount) {
      const facotr = (new BigNumber(10)).pow(token.decimals);
      if (!(rawAmount instanceof BigNumber)) {
        rawAmount = new BigNumber(rawAmount);
      }
      return rawAmount.div(facotr);
    } else {
      return new BigNumber(0);
    }
  }

  toRawAmount(token, amount, fixedMethod = 'truncated') {
    if (amount) {
      const facotr = (new BigNumber(10)).pow(token.decimals);
      if (!(amount instanceof BigNumber)) {
        amount = new BigNumber(amount);
      }
      return amount.mul(facotr)[fixedMethod]();
    } else {
      return new BigNumber(0);
    }
  }
}

module.exports = TokenManager;


src/joyso.js
const rp = require('request-promise');
const ethUtil = require('ethereumjs-util');
const BigNumber = require('bignumber.js');
const _ = require('lodash');
const ActionCable = require('./actioncable');
const System = require('./system');
const TokenManager = require('./token_manager');
const Balances = require('./balances');
const Orders = require('./orders');
const OrderBook = require('./order-book');
const Trades = require('./trades');
const MyTrades = require('./my-trades');
const Funds = require('./funds');
const Account = require('./account');

BigNumber.config({ DECIMAL_PLACES: 36 });

const ACCESS_TOKEN_TIME = 600000;
const MAX_FEE_PRICE = new BigNumber('100000000');
const TOKEN_MAX_FEE_PRICE = new BigNumber('1000000000000000000000000000');
const keys = {};

class Joyso {
  constructor(options = {}) {
    this.host = options.host || 'tron.joyso.io';
    this.ssl = options.ssl === undefined ? true : options.ssl;
    let key = options.key;
    if (key.indexOf('0x') !== 0) {
      key = `0x${key}`;
    }
    this.keyIndex = Object.keys(keys).length;
    keys[this.keyIndex] = key;
    this.address = `0x${ethUtil.privateToAddress(key).toString('hex')}`;
    this.orderBooks = {};
    this.trades = {};
    this.hashTable = {};
  }

  async connect() {
    if (this.connected) {
      return;
    }
    await this.updateAccessToken();
    this.cable = ActionCable.createConsumer(this.wsUrl, this.origin);
    this.system = new System(this);
    this.system.once('update', json => {
      this.tokenManager = new TokenManager(this, json);
      this.tokenManager.subscribe();
    });
    await this.system.connect();
    this.account = new Account(this, this.address);
    await this.account.connect();
    this.balances = new Balances({ client: this, address: this.address });
    await this.balances.subscribe();
    this.orders = new Orders({
      client: this,
      address: this.address
    });
    this.myTrades = new MyTrades({
      client: this,
      address: this.address
    });
    this.funds = new Funds({
      client: this,
      address: this.address
    });
    this.connected = true;
  }

  async disconnect() {
    this.cable.disconnect();
    clearTimeout(this.timer);
    this.connected = false;
    delete this.accessToken;
  }

  async updateAccessToken(catchError = false) {
    const nonce = Math.floor(Date.now() / 1000);
    const raw = `Signing this message proves to JOYSO you are in control of your account without giving JOYSO access to any sensitive information. Message ID: ${nonce}`;
    const vrs = this.sign(new Buffer(raw, 'utf8'));
    try {
      const r = await this.request('accounts', {
        method: 'POST',
        data: Object.assign({
          user: this.address.substr(2),
          nonce: nonce,
          version: 2,
          locale: 'en'
        }, vrs)
      });
      this.accessToken = r.access_token;
    } catch (e) {
      if (!catchError) {
        throw e;
      } else if (e.statusCode === 400) {
        throw new Error('Failed to update access token.');
      }
    } finally {
      this.timer = setTimeout(() => this.updateAccessToken(true), ACCESS_TOKEN_TIME);
    }
  }

  async withdraw(options) {
    const withdraw = this.createWithdraw(options);
    try {
      const r = await this.request('withdraw_queues', {
        method: 'POST',
        data: withdraw,
        auth: true
      });
    } catch (e) {
      if (!options.retry && e.statusCode === 400 && e.error.fee_changed) {
        options.retry = true;
        await this.tokenManager.refresh();
        return this.withdraw(options);
      }
      throw e;
    }
  }

  repayWithdrawFee(token) {
    if (this.system.advanceable && this.account.advanceReal !== 0 && this.account.advanceInOrder === 0) {
      const ratio = this.tokenManager.trx.withdrawFee.div(token.withdrawFee);
      return new BigNumber(this.account.advanceReal).div(ratio).truncated().add(token.withdrawFee);
    } else {
      return token.withdrawFee;
    }
  }

  createWithdraw({ token, amount, fee }) {
    this.validateWithdraw(amount, fee);
    let tokenFee, paymentMethod;
    if (fee === 'token') {
      tokenFee = this.tokenManager.symbolMap[token];
      paymentMethod = 2;
    } else if (fee === 'joy') {
      tokenFee = this.tokenManager.joy;
      paymentMethod = 1;
    } else {
      tokenFee = this.tokenManager.trx;
      paymentMethod = 0;
      fee = 'eth';
    }
    token = this.tokenManager.symbolMap[token];
    if (!token) {
      throw new Error('invalid token');
    }
    const withdrawFee = this.repayWithdrawFee(tokenFee);
    let rawAmount = this.tokenManager.toRawAmount(token, amount);
    if (token === tokenFee) {
      rawAmount = rawAmount.sub(withdrawFee);
    }
    const timestamp = Date.now();
    const nonce = Math.floor(timestamp / 1000);
    let data = _.padStart(nonce.toString(16), 8, '0');
    data += '000000000000000';
    data += paymentMethod;

    let input = this.system.contract.substr(2);
    input += _.padStart(rawAmount.toString(16), 64, '0');
    input += _.padStart(withdrawFee.toString(16), 64, '0');
    input += data;
    input += token.address.substr(2);
    const hash = ethUtil.keccak256(new Buffer(input, 'hex'));
    const vrs = this.sign(hash);

    return Object.assign({
      nonce,
      contract: this.system.contract.substr(2),
      amount: rawAmount.toString(10),
      fee: withdrawFee.toString(10),
      user: this.address.substr(2),
      token: token.address.substr(2),
      payment_method: fee
    }, vrs)
  }

  async buy(options) {
    options.side = 'buy';
    return this.trade(options);
  }

  async sell(options) {
    options.side = 'sell';
    return this.trade(options);
  }

  async trade(options) {
    const order = this.createOrder(options);
    try {
      const hash = order.hash;
      delete order.hash;
      const r = await this.request('orders', {
        method: 'POST',
        data: order,
        auth: true
      });
      this.updateHashTable(order.nonce, hash);
      order.id = r.order.id;
      order.status = r.order.status;
      order.amount_fill = r.order.amount_fill;
      return this.orders.convert([order])[0];
    } catch (e) {
      if (!options.retry && e.statusCode === 400 && e.error.fee_changed) {
        await this.tokenManager.refresh();
        options.retry = true;
        return this.trade(options);
      }
      throw e;
    }
  }

  async getMyTrades({ from, to, quote, base, side, before, limit } = {}) {
    const options = {};
    quote = this.tokenManager.symbolMap[quote];
    if (quote) {
      if (this.tokenManager.quotes.find(t => t === quote)) {
        options.token_base = quote.address.substr(2);
      } else {
        throw new Error('invalid quote');
      }
    }
    base = this.tokenManager.symbolMap[base];
    if (base) {
      options.token_target = base.address.substr(2);
    }
    if (side === 'buy') {
      options.is_buy = true;
    } else if (side === 'sell') {
      options.is_buy = false
    }
    if (from) {
      options.from = from;
    }
    if (to) {
      options.to = to;
    }
    if (before) {
      options.before = before;
    }
    if (limit) {
      options.limit = limit;
    }
    options.user = this.address.substr(2);
    const json = await this.request('trades/history', { data: options });
    return this.myTrades.convert(json.trades);
  }

  toAmountByPrice(token, baseAmount, price, method) {
    const amount = this.tokenManager.toRawAmount(token, baseAmount.mul(price), method);
    return this.tokenManager.toAmount(token, amount);
  }

  toPrice(baseAmount, quoteAmount) {
    return quoteAmount.div(baseAmount).round(9);
  }

  validateAmount(amount) {
    const lte = amount instanceof BigNumber ? amount.lte(0) : amount <= 0;;
    if (lte) {
      throw new Error('invalid amount');
    }
  }

  validateWithdraw(amount, fee) {
    this.validateAmount(amount);
    if (fee !== 'trx' && fee !== 'joy' && fee !== 'token') {
      throw new Error('invalid fee');
    }
  }

  validateOrder(price, amount, side, quote) {
    this.validateAmount(amount);
    const precision = new BigNumber(10).pow(quote.precision);
    const v = new BigNumber(price).mul(precision);
    if (!v.truncated().equals(v)) {
      throw new Error('invalid price');
    }
    if (side !== 'buy' && side !== 'sell') {
      throw new Error('invalid side');
    }
  }

  validatePair(base, quote) {
    if (!base || !quote || !this.tokenManager.quotes.find(t => t === quote)) {
      throw new Error('invalid pair');
    }
  }

  repayGasFee(token) {
    const ratio = new BigNumber(this.tokenManager.trx.gasFee).div(token.gasFee);
    return new BigNumber(this.account.advanceReal).div(ratio).truncated().add(token.gasFee);
  }

  receivableGasFee(side, feeByJoy, tokenFee, quote) {
    let gasFee = quote.gasFee;
    let baseBalance = this.balances.balances[quote.symbol];
    baseBalance = this.tokenManager.toRawAmount(quote, baseBalance && baseBalance.available || 0);
    if (this.system.advanceable && side !== 'buy' && !feeByJoy && gasFee.gt(baseBalance)
      && this.account.advanceReal === 0 && this.account.advanceInOrder === 0
    ) {
      return new BigNumber(0);
    } else if (this.account.advanceReal !== 0 && this.account.advanceInOrder === 0) {
      return this.repayGasFee(tokenFee);
    } else {
      return tokenFee.gasFee;
    }
  }

  createOrder({ pair, price, amount, feeByJoy, side }) {
    const [base, quote]= this.tokenManager.getPair(pair);
    this.validatePair(base, quote);
    this.validateOrder(price, amount, side, quote);
    let baseAmount = new BigNumber(amount);
    let method = side === 'buy' ? 'ceil' : 'floor';
    let quoteAmount = this.toAmountByPrice(quote, baseAmount, price, method);
    if (!this.toPrice(baseAmount, quoteAmount).equals(price)) {
      method = method === 'floor' ? 'ceil' : 'floor';
      quoteAmount = this.toAmountByPrice(quote, baseAmount, price, method);
      if (!this.toPrice(baseAmount, quoteAmount).equals(price)) {
        throw new Error('invalid amount, too small');
      }
    }
    baseAmount = this.tokenManager.toRawAmount(base, baseAmount);
    quoteAmount = this.tokenManager.toRawAmount(quote, quoteAmount);

    let takerFee, makerFee, tokenFee, feePrice, custom = false;
    if (base.taker_fee && base.maker_fee) {
      takerFee = base.taker_fee;
      makerFee = base.maker_fee;
      custom = true;
    } else {
      takerFee = this.system.takerFee;
      makerFee = this.system.makerFee;
    }

    if (feeByJoy) {
      tokenFee = this.tokenManager.joy;
      if (!custom) {
        takerFee = Math.floor(takerFee / 2);
        makerFee = Math.floor(makerFee / 2);
      }
      feePrice = tokenFee.price;
      let maxFeePrice, offset;
      if (quote === this.tokenManager.trx) {
        maxFeePrice = MAX_FEE_PRICE;
        offset = new BigNumber('10000000');
      } else {
        const priceToTrx = quote.price;
        if (feePrice && priceToTrx) {
          const decimalsOffset = new BigNumber(10).pow(tokenFee.decimals - quote.decimals);
          offset = new BigNumber('1000000000000').div(decimalsOffset);
          feePrice = new BigNumber(feePrice).div(priceToTrx);
        } else {
          throw new Error('fee price invalid');
        }
        maxFeePrice = TOKEN_MAX_FEE_PRICE;
      }
      feePrice = feePrice.mul(offset).truncated();
      if (feePrice.gt(maxFeePrice)) {
        feePrice = maxFeePrice;
      } else if (feePrice.lt(1)) {
        feePrice = new BigNumber(1);
      }
    } else {
      tokenFee = quote;
      feePrice = 0;
    }
    const gasFee = this.receivableGasFee(side, feeByJoy, tokenFee, quote);

    let amountSell, amountBuy, tokenSell, tokenBuy;
    if (side === 'buy') {
      amountSell = quoteAmount;
      amountBuy = baseAmount;
      tokenSell = quote.address;
      tokenBuy = base.address;
    } else {
      amountSell = baseAmount;
      amountBuy = quoteAmount;
      tokenSell = base.address;
      tokenBuy = quote.address;
    }

    const createHash = (nonce) => {
      let input;
      if (quote === this.tokenManager.trx) {
        let data = _.padStart(nonce.toString(16), 8, '0');
        data += _.padStart(takerFee.toString(16), 4, '0');
        data += _.padStart(makerFee.toString(16), 4, '0');
        data += _.padStart(feePrice.toString(16), 7, '0');
        data += side === 'buy' ? '1' : '0';

        input = this.system.contract.substr(2);
        input += _.padStart(amountSell.toString(16), 64, '0');
        input += _.padStart(amountBuy.toString(16), 64, '0');
        input += _.padStart(gasFee.toString(16), 64, '0');
        input += data;
        input += base.address.substr(2);
      } else {
        let data = _.padStart(nonce.toString(16), 8, '0');
        data += _.padStart(takerFee.toString(16), 4, '0');
        data += _.padStart(makerFee.toString(16), 4, '0');
        data += '0000000';
        data += side === 'buy' ? '1' : '0';

        input = this.system.contract.substr(2);
        input += _.padStart(amountSell.toString(16), 64, '0');
        input += _.padStart(amountBuy.toString(16), 64, '0');
        input += _.padStart(gasFee.toString(16), 64, '0');
        input += data;
        input += base.address.substr(2);
        input += quote.address.substr(2);
        input += _.padStart(feePrice.toString(16), 64, '0');
      }
      return ethUtil.keccak256(new Buffer(input, 'hex'));
    };

    const timestamp = Date.now();
    const nonce = Math.floor(timestamp / 1000);
    let hash = createHash(nonce);
    let retry = 0;
    while (this.hashTable[nonce] && this.hashTable[nonce][hash] && retry < 60) {
      hash = createHash(nonce + ++retry);
    }
    const vrs = this.sign(hash);
    return Object.assign({
      nonce,
      maker_fee: makerFee,
      taker_fee: takerFee,
      fee_price: feePrice.toString(10),
      token_sell: tokenSell.substr(2),
      token_buy: tokenBuy.substr(2),
      user: this.address.substr(2),
      contract: this.system.contract.substr(2),
      gas_fee: gasFee.toString(10),
      amount_sell: amountSell.toString(10),
      amount_buy: amountBuy.toString(10),
      payment_method: feeByJoy ? 'joy' : 'base',
      hash: hash.toString('hex'),
      is_buy: side === 'buy'
    }, vrs);
  }

  cancel(orderId) {
    return this.request(`orders/${orderId}`, {
      method: 'DELETE',
      auth: true
    });
  }

  subscribeOrderBook(pair, callback) {
    if (!this.connected) {
      throw new Error('client is not connected');
    }
    const [base, quote] = this.tokenManager.getPair(pair);
    this.validatePair(base, quote);
    if (this.orderBooks[pair]) {
      this.orderBooks[pair].unsubscribe();
    }
    this.orderBooks[pair] = new OrderBook({
      base, quote,
      client: this,
      onReceived: callback,
      onUnsubscribe: () => delete this.orderBooks[pair]
    });
    this.orderBooks[pair].subscribe();
    return this.orderBooks[pair];
  }

  subscribeTrades(pair, callback) {
    if (!this.connected) {
      throw new Error('client is not connected');
    }
    const [base, quote] = this.tokenManager.getPair(pair);
    this.validatePair(base, quote);
    if (this.trades[pair]) {
      this.trades[pair].unsubscribe();
    }
    this.trades[pair] = new Trades({
      base, quote,
      client: this,
      onReceived: callback,
      onUnsubscribe: () => delete this.trades[pair]
    });
    this.trades[pair].subscribe();
    return this.trades[pair];
  }

  subscribeMyTrades(callback) {
    if (!this.connected) {
      throw new Error('client is not connected');
    }
    this.myTrades.onReceived = callback;
    if (this.myTrades.cable) {
      this.myTrades.unsubscribe();
    }
    this.myTrades.subscribe();
    return this.myTrades;
  }

  subscribeOrders(callback) {
    if (!this.connected) {
      throw new Error('client is not connected');
    }
    this.orders.onReceived = callback;
    if (this.orders.cable) {
      this.orders.unsubscribe();
    }
    this.orders.subscribe();
    return this.orders;
  }

  subscribeBalances(callback) {
    if (!this.connected) {
      throw new Error('client is not connected');
    }
    this.balances.onReceived = callback;
    this.balances.onReceived(this.balances.balances);
    return this.balances;
  }

  subscribeFunds(callback) {
    if (!this.connected) {
      throw new Error('client is not connected');
    }
    this.funds.onReceived = callback;
    if (this.funds.cable) {
      this.funds.unsubscribe();
    }
    this.funds.subscribe();
    return this.funds;
  }

  sign(message) {
    const result = ethUtil.keccak256(Buffer.concat([new Buffer('\x19TRON Signed Message:\n32'), message]));
    const vrs = ethUtil.ecsign(result, ethUtil.toBuffer(keys[this.keyIndex]));
    return {
      v: vrs.v,
      r: vrs.r.toString('hex'),
      s: vrs.s.toString('hex'),
    };
  }

  updateHashTable(nonce, hash) {
    if (!this.hashTable[nonce]) {
      this.hashTable[nonce] = {};
    }
    this.hashTable[nonce][hash] = true;
    const now = Math.floor(Date.now() / 1000);
    Object.keys(this.hashTable).forEach(k => {
      if (k < now) {
        delete this.hashTable[k];
      }
    });
  }

  get origin() {
    const protocol = this.ssl ? 'https' : 'http'
    return `${protocol}://${this.host}`;
  }

  get apiUrl() {
    return `${this.origin}/api/v1`;
  }

  get wsUrl() {
    const protocol = this.ssl ? 'wss' : 'ws'
    return `${protocol}://${this.host}/cable`;
  }

  request(path, options = {}) {
    let { headers, method, auth, data = {} } = options;
    delete options.data;
    delete options.auth;

    if (auth) {
      headers = headers || {};
      headers['Authorization'] = `Bearer ${this.accessToken}`;
      options.headers = headers;
    }

    const attr = (method || 'GET').toUpperCase() == 'GET' ? 'qs' : 'body';
    options[attr] = options[attr] || {};
    Object.assign(options[attr], data);

    options = Object.assign({
      uri: `${this.apiUrl}/${path}`,
      json: true
    }, options);
    return rp(options);
  }
}

module.exports = Joyso;
